From: Bartlomiej Dudek <bartlomiej.dudek@intel.com>
Date: Fri, 4 Aug 2017 13:52:32 -0700
Subject: IB/hfi1: Use host_link_state to read state when DC is shut down
Patch-mainline: v4.14-rc1
Git-commit: 64a296f579303322ebec9edae09cf87240b1ad78
References: bsc#1060463 FATE#323043

When DC is shut down (by e.g.  disconnecting the cable), the
driver should use host_link_state to get port's current
physical state. This is due to the fact that physical state
is read from DC's CSRs and when DC is shut down and state is
changed, its registers are not impacted.

Reviewed-by: Jakub Byczkowski <jakub.byczkowski@intel.com>
Signed-off-by: Bartlomiej Dudek <bartlomiej.dudek@intel.com>
Signed-off-by: Dennis Dalessandro <dennis.dalessandro@intel.com>
Signed-off-by: Doug Ledford <dledford@redhat.com>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/infiniband/hw/hfi1/hfi.h |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/infiniband/hw/hfi1/hfi.h
+++ b/drivers/infiniband/hw/hfi1/hfi.h
@@ -1298,6 +1298,13 @@ int hfi1_reset_device(int);
 static inline u32 driver_pstate(struct hfi1_pportdata *ppd)
 {
 	/*
+	 * When DC is shut down and state is changed, its CSRs are not
+	 * impacted, therefore host_link_state should be used to get
+	 * current physical state.
+	 */
+	if (ppd->dd->dc_shutdown)
+		return driver_physical_state(ppd);
+	/*
 	 * The driver does some processing from the time the physical
 	 * link state is at LINKUP to the time the SM can be notified
 	 * as such. Return IB_PORTPHYSSTATE_TRAINING until the software
